src/execution/sql/thread_state_container.cpp

```diff
@@ -83,35 +87,52 @@ void ThreadStateContainer::Reset(const std::size_t state_size, const ThreadState
 }
 
 byte *ThreadStateContainer::AccessCurrentThreadState() {
-  auto &tls_handle = impl_->states_.local();
-  return tls_handle.State();
+  common::SpinLatch::ScopedSpinLatch guard(&impl_->states_latch_);
+  auto iter = impl_->states_.find(std::this_thread::get_id());
+  byte *state = nullptr;
```

Comment:
This pattern will perform two lookups into the hash table, which might explain the performance you're seeing. What you want to do is:
```
auto &tls_handle = impl_->states_[std::this_thread::get_id()];
if (!tls_handle) {
  tls_handle = std::make_unique<TLSHandle>(this);
}
return tls_handle->State();
```

Check the syntax, but the above should be faster. You should also consider `std::unordered_map::reserve()` at `ThreadStateContainer` construction time to pre-allocate room to avoid resizing while holding the latch here.

Reply:
In our code, find is happening just once. If the entry is not found, it is being inserted into the map for future look ups.

Reply:
Not quite: the version above probes the map **exactly** once per lookup, regardless of whether it's a hit or miss. Your version accesses it twice on miss. This means double the hashing and double the memory references, all under a latch.

This function is called only once-per-pipeline 90% of the time (for TPCC it's probably 100%). This means calling threads incur the full double-map-access since the states are reset in each pipeline.

Reply:
The initial state of the map is empty. With the code above, every call to AccessCurrentThread would result in thread id not being found and just returning state for a new TLSHandle object. I am not sure the map will even be used in that case. Are you thinking of populating the map in the constructor of the ThreadStateContainer class instead of in the AccessCurrentThreadState function?

Reply:
Take a look at the docs for [`std::unordered_map::operator[]`](https://en.cppreference.com/w/cpp/container/unordered_map/operator_at). It inserts the key and default initializes the value if the key is new. In both cases, it returns a reference to the value. The version above checks if the returned value is `nullptr` (indicating a miss) and move-assigns the value to a new state. This achieves the single map access guarantee.

Reply:
For `std::unordered_map::reserve()` shall I pass double the number of cores? We did run 36 threads on an 18 core machine.

Reply:
Try three versions: no reserve, reserve 1x core count, and reserve 2x core count, all with `std::unordered_map<std::thread::id, std::unique_ptr<>>` to see performance.

Ideally, the `ThreadStateContainer` constructor should accept an expected thread count. But, the scope of that change is much larger.